pytorch_lightning/profiler/pytorch.py

```diff
@@ -448,68 +613,242 @@ def on_trace_ready(profiler):
+
+        step_num = self._action_step_num[action_name]
+        if step_num == self._wait_step:
+            # warm up
+            self._transit_profiler(KinetoProfilerState.WARMUP)
+        elif step_num >= self._wait_step + self._warmup_step and step_num < self.profile_steps:
+            # begin profile
+            self._transit_profiler(KinetoProfilerState.START)
+
+    def _stop_action(self, action_name: str) -> None:
+        if self.profiler is not None and any(action_name.endswith(func) for func in self.STEP_FUNCTIONS):
+            # Save the first action name in order to save the profile data
+            # when destroying profiler like what the original on_trace_ready did.
+            if self._start_action_name is None:
+                self._start_action_name = action_name
+
+            if self._action_step_num[action_name] == self.profile_steps:
+                self._transit_profiler(KinetoProfilerState.NONE)
+                if self.dirpath is not None:
+                    self._save_result(action_name)
+                else:
+                    rank_zero_warn("The PyTorchProfiler failed to export trace as `dirpath` is None")
+
+    def _delete_profiler(self) -> None:
         if self.profiler is not None:
-            self.profiler.__exit__(None, None, None)
+            if not _KINETO_AVAILABLE or self._emit_nvtx:
+                self.profiler.__exit__(None, None, None)
+            else:
+                # At the moment, the profiler should already be stoped.
+                if torch.autograd._profiler_enabled():
+                    self._transit_profiler(KinetoProfilerState.NONE)
+                    self._save_result(self._start_action_name)
             self._cache_functions_events()
             self.profiler = None
 
-        if self._schedule is not None:
-            self._schedule.reset()
+    def _save_result(self, action_name: str) -> None:
+        os.makedirs(self.dirpath, exist_ok=True)
+        if self._export_to_chrome:
+            path = os.path.join(
+                self.dirpath,
+                "{}.{}.pt.trace.json".format(
+                    self._prepare_filename(action_name=action_name, extension=""), int(time.time() * 1000)
+                ),
+            )
+            self.profiler.export_chrome_trace(path)
+        if self._export_to_flame_graph:
+            path = os.path.join(self.dirpath, self._prepare_filename(action_name=action_name, extension=".stack"))
+            self.profiler.export_stacks(path, metric=self._metric)
 
-        if self._parent_profiler is not None:
-            self._parent_profiler.__exit__(None, None, None)
-            self._parent_profiler = None
+    @property
+    def profile_steps(self):
+        return self._wait_step + self._warmup_step + self._active_step
 
-        if self._register is not None:
-            self._register.__exit__(None, None, None)
-            self._register = None
+    def get_total_steps(self, action_name) -> int:
```

Comment:
Can you extract this logic into a protected function and use it both here and in `_total_steps`?